gobblin-yarn/src/main/java/gobblin/yarn/GobblinYarnAppLauncher.java

```diff
@@ -314,6 +344,33 @@ public void handleApplicationReportArrivalEvent(ApplicationReportArrivalEvent ap
         LOGGER.error("Failed to close the " + GobblinYarnAppLauncher.class.getSimpleName(), ioe);
       } catch (TimeoutException te) {
         LOGGER.error("Timeout in stopping the service manager", te);
+      } finally {
+        if (this.emailNotificationOnShutdown) {
+          sendEmailOnShutdown(Optional.of(applicationReport));
+        }
+      }
+    }
+  }
+
+  @Subscribe
+  public void handleGetApplicationReportFailureEvent(
```

Comment:
If this method could be called concurrently, it should probably be synchronized to avoid sending multiple email notifications. If not, it isn't necessary to use AtomicInteger.